onnxruntime/test/ir/onnx_model_test.cc

```diff
@@ -187,5 +189,152 @@ TEST_F(ONNXModelsTest, TestModelsWithAnOpContainingAFunctionBody) {
+  EXPECT_TRUE(parser.EndOfInput());
+
+  // Build first function proto
+  auto* function_proto = model_proto.mutable_functions()->Add();
+  auto func_body_nodes = FunctionBodyHelper::BuildNodes(
+      {// nodes: {outputs, op, inputs, attributes, domain}
+       FunctionBodyHelper::Const<float>("Q_Min", 0.f),
+       FunctionBodyHelper::Const<float>("Q_Max", 255.f),
+       {{"X_Min"}, "ReduceMin", {"x"}, {MakeAttribute("keepdims", int64_t(0))}},
+       {{"X_Max"}, "ReduceMax", {"x"}, {MakeAttribute("keepdims", int64_t(0))}},
+       {{"X_Range"}, "Sub", {"X_Max", "X_Min"}},
+       {{"Scale"}, "Div", {"X_Range", "Q_Max"}},
+       {{"ZeroPoint_FP"}, "Sub", {"Q_Min", "Scale"}},
+       {{"Zeropoint"}, "Cast", {"ZeroPoint_FP"}, {MakeAttribute("to", int64_t(2))}},
+       {{"y"}, "QuantizeLinear", {"x", "Scale", "Zeropoint"}}});
+
+  for (const auto& node : func_body_nodes) {
+    auto new_node = function_proto->add_node();
+    new_node->CopyFrom(node);
+  }
+
+  function_proto->set_name("foo");
+  function_proto->set_domain("custom_domainB");
+  function_proto->set_doc_string("Test function proto");
+  function_proto->add_input("x");
+  function_proto->add_output("y");
+
+  std::unordered_map<std::string, int> opset_imports({{"", 13}});
+  for (auto& opset_import : opset_imports) {
+    auto* func_opset_import = function_proto->mutable_opset_import()->Add();
+    func_opset_import->set_domain(opset_import.first);
+    func_opset_import->set_version(opset_import.second);
+  }
+
+  // Build second function proto
+  // intentionally using same function name to test 
+  // that domainA:name and domainB:name are allowed.
+  function_proto = model_proto.mutable_functions()->Add();
+  func_body_nodes = FunctionBodyHelper::BuildNodes(
+      {// nodes: {outputs, op, inputs, attributes, domain}
+       {{"y"}, "foo", {"x"}, {}, "custom_domainB"},
+       {{"c"}, "Identity", {"y"}}});
+
+  for (const auto& node : func_body_nodes) {
+    auto new_node = function_proto->add_node();
+    new_node->CopyFrom(node);
+  }
+
+  function_proto->set_name("foo");
+  function_proto->set_domain("custom_domainA");
+  function_proto->set_doc_string("Test function proto");
+  function_proto->add_input("x");
+  function_proto->add_output("c");
+
+  std::unordered_map<std::string, int> func2_opset_imports({{"", 13}, {"custom_domainB", 1}});
+  for (auto& opset_import : func2_opset_imports) {
+    auto* func_opset_import = function_proto->mutable_opset_import()->Add();
+    func_opset_import->set_domain(opset_import.first);
+    func_opset_import->set_version(opset_import.second);
+  }
+
+  std::shared_ptr<Model> model;
+  std::shared_ptr<onnxruntime::OnnxRuntimeOpSchemaRegistry> registry = std::make_shared<OnnxRuntimeOpSchemaRegistry>();
+  std::list<std::shared_ptr<IOnnxRuntimeOpSchemaCollection>> regs = {registry};
+  ASSERT_STATUS_OK(Model::Load(std::move(model_proto), model, &regs, *(DefaultLoggingManager().CreateLogger("GraphTest"))));
+}
 }  // namespace test
```

Comment:
Is there a test that uses the same function multiple times in a graph?

Reply:
added a new test: TestNestedModelLocalFunctionsWithMultipleReferences